### ISSUE TITLE
SF Remoting V1/V2 causes indefinite hanging.

### ISSUE THREAD

**Issue by NArnott** [Azure/service-fabric-issues#954]
When setting up a service fabric remoting transport using V2, if I forget to set the endpoint name to ServiceEndpoint**V2**, the client will hang forever and no exceptions or logs will be created. I know V1 and V2 are incompatible, but is there a way to get them to throw an exception somewhere if the two different versions attempt to communicate?

**Comment by suchiagicha** [comment]
@NArnott 
No. We consider both cases ( failover scenario and incompatible version ) as communication failure and retry on those.  Its like client doesn't know if service is not there or you are talking to wrong service endpoint.

**Comment by NArnott** [comment]
I think a max retry option even for transient errors would be helpful. How else could we maintain an SLA using remoting? I don't like the idea of these calls hanging (for whatever reason) for days, weeks, etc. If not by default, it makes sense to be able to set an an upper cap so it can finally error out and let us log it so we can't find the issue.

**Comment by suchiagicha** [comment]
You can have cancellation token as a parameter to your remote method in your remoting Interface , That way it wont get hang in these kind of scenarios. 
E,g You can also pass timeout to cancellation token.

**Comment by NArnott** [comment]
@suchiagicha That would be very helpful. Is there any documentation on this? Maybe some examples?

**Comment by suchiagicha** [comment]
Below Api has a cancellation token as parameter. Here is an article on how to use cancellation token
https://msdn.microsoft.com/en-us/library/system.threading.cancellationtoken(v=vs.110).aspx


public interface IMyService : IService
{

  Task<string> HelloWorldAsyncWithToken(CancellationToken cancellationtoken);
}

**Comment by NArnott** [comment]
@suchiagicha I am familiar with cancellation tokens. I was hoping for some documentation on their use with SF remoting. The example you gave is simple enough, but what about when other parameters are used? If the token is canceled, is that signal passed to the service side as well? I'm hoping to find documentation that explains all that.

**Comment by NArnott** [comment]
@suchiagicha Is there a way to hook into the pipeline directly? For example, to set a default max timeout of 30 seconds without having to add that argument to all the remoting methods?

**Comment by suchiagicha** [comment]
@NArnott 
No. We recommend users to use cancellationToken in there Api . In case of long-running operation , cancellation token helps you to cancel the operation on service side if needed, Retry setting wont in this case.

**Comment by MarekSalat** [comment]
@suchiagicha  hello, is there any update on this topic?

**Comment by suchiagicha** [comment]
@MarekSalat
You could use cancellation token in your API to cancel the call and then it throws OperationCancelledException.
It hangs as we consider this scenario as wrong endpoint for the connecting service. We re-resolve and send request again.

**Comment by WhitWaldo** [comment]
@suchiagicha Was the documentation ever updated to reflect this example?
